### PR TITLE
updating new helper

### PR DESCRIPTION
**Pull request by iruzevic**
- updating manifest layout for custom variable key
- new helper to map responsive attributes in single array

### CODE REVIEW

File: src/Helpers/Components.php [hunk 1]
```diff
@@ -258,6 +258,38 @@ public static function checkAttr(string $key, array $attributes, array $manifest
...
+
+		foreach ($manifest['responsiveAttributes'][$keyName] as $key => $value) {
+			$output[$key] = self::checkAttr($value, $attributes, $manifest, $componentName);
+		}
+
+		return $output;
+	}
+
 	/**
 	 * Return BEM selector for html class and check if Condition part is set.
```
Comment (goranalkovic-infinum) [inline src/Helpers/Components.php:294]:
```suggestion
	 * Return a BEM class selector and check if Condition part is set.
```

File: src/Helpers/Components.php [hunk 2]
```diff
@@ -258,6 +258,38 @@ public static function checkAttr(string $key, array $attributes, array $manifest
 		}
 	}
 
+	/**
+	 * Map and check attributes for responsive object.
+	 *
+	 * @param string $keyName Key name to find in responsiveAttributes object.
```
Comment (goranalkovic-infinum) [inline src/Helpers/Components.php:264]:
```suggestion
	 * @param string $keyName Key name to find in the responsiveAttributes object.
```

File: src/Helpers/Components.php [hunk 3]
```diff
@@ -461,6 +499,27 @@ function ($item) use ($attributes, $key) {
 		";
 	}
 
+	/**
+	 * Internal helper to loop Css Variables from object.
```
Comment (goranalkovic-infinum) [inline src/Helpers/Components.php:503]:
```suggestion
	 * Internal helper to loop CSS variables from an object.
```

File: src/Helpers/Components.php [hunk 4]
```diff
@@ -461,6 +499,27 @@ function ($item) use ($attributes, $key) {
...
+	 *
+	 * @param array  $arrayList Array list of css variables.
```
Comment (goranalkovic-infinum) [inline src/Helpers/Components.php:505]:
```suggestion
	 * @param array  $arrayList List of CSS variables.
```

File: src/Helpers/Components.php [hunk 5]
```diff
@@ -461,6 +499,27 @@ function ($item) use ($attributes, $key) {
...
+	 * @param string $attributeKey Attribute key to append to output variable name.
```
Comment (goranalkovic-infinum) [inline src/Helpers/Components.php:506]:
```suggestion
	 * @param string $attributeKey Attribute key to append to the output variable name.
```

File: tests/Helpers/ComponentHelpersTest.php [hunk 6]
```diff
@@ -243,6 +243,49 @@ function ($argument) {
...
+	$attributes = [
+		'headingContentSpacingLarge' => '10',
+		'headingContentSpacingDesktop' => '5',
+		'headingContentSpacingTablet' => '3',
+		'headingContentSpacingMobile' => '1',
+	];
+
+	$results = Components::checkAttrResponsive('headingContentSpacing', $attributes, $manifest);
+
+	$this->assertIsArray($results, 'Result should be a array');
```
Comment (goranalkovic-infinum) [inline tests/Helpers/ComponentHelpersTest.php:260]:
```suggestion
	$this->assertIsArray($results, 'Result should be an array');
```

File: tests/Helpers/ComponentHelpersTest.php [hunk 7]
```diff
@@ -243,6 +243,49 @@ function ($argument) {
 	Components::checkAttr('bla', $attributes, $manifest, 'button');
 })->throws(\Exception::class, 'bla key does not exist in the button component. Please check your implementation.');
 
+/**
+ * Components::checkAttrResponsive tests
+ */
+test('Asserts that checkAttrResponsive returns correct output.', function () {
```
Comment (goranalkovic-infinum) [inline tests/Helpers/ComponentHelpersTest.php:249]:
```suggestion
test('Asserts that checkAttrResponsive returns the correct output.', function () {
```

File: tests/Helpers/ComponentHelpersTest.php [hunk 8]
```diff
@@ -243,6 +243,49 @@ function ($argument) {
...
+	$this->assertEquals($results['large'], '10');
+});
+
+test('Asserts that checkAttrResponsive returns empty values if attribute is not provided.', function () {
+	$manifest = Components::getManifest(dirname(__FILE__, 2) . '/data/src/Blocks/components/heading/');
+	$attributes = [];
+
+	$results = Components::checkAttrResponsive('headingContentSpacing', $attributes, $manifest);
+
+	$this->assertIsArray($results, 'Result should be a array');
```
Comment (goranalkovic-infinum) [inline tests/Helpers/ComponentHelpersTest.php:271]:
```suggestion
	$this->assertIsArray($results, 'Result should be an array');
```

File: tests/Helpers/ComponentHelpersTest.php [hunk 9]
```diff
@@ -243,6 +243,49 @@ function ($argument) {
...
+	$this->assertArrayHasKey('large', $results);
+	$this->assertEquals($results['large'], '');
+});
+
+test('Asserts that checkAttrResponsive throws error if responsiveAttribute key is missing.', function () {
+	$manifest = Components::getManifest(dirname(__FILE__, 2) . '/data/src/Blocks/components/button/');
+	$attributes = [];
+
+	Components::checkAttrResponsive('headingContentSpacing', $attributes, $manifest, 'button');
+})->throws(\Exception::class, 'It looks like you are missing responsiveAttributes key in your button manifest.');
```
Comment (goranalkovic-infinum) [inline tests/Helpers/ComponentHelpersTest.php:281]:
```suggestion
})->throws(\Exception::class, 'It looks like you are missing the responsiveAttributes key in your button manifest.');
```

File: tests/Helpers/ComponentHelpersTest.php [hunk 10]
```diff
@@ -243,6 +243,49 @@ function ($argument) {
...
+
+test('Asserts that checkAttrResponsive throws error if keyName key is missing responsiveAttributes array.', function () {
+	$manifest = Components::getManifest(dirname(__FILE__, 2) . '/data/src/Blocks/components/heading/');
+	$attributes = [];
+
+	Components::checkAttrResponsive('testAttribtue', $attributes, $manifest, 'button');
+})->throws(\Exception::class, 'It looks like you are missing testAttribtue key in your manifest responsiveAttributes array.');
```
Comment (goranalkovic-infinum) [inline tests/Helpers/ComponentHelpersTest.php:288]:
```suggestion
})->throws(\Exception::class, 'It looks like you are the missing testAttribtue key in your manifest responsiveAttributes array.');
```

File: tests/Helpers/ComponentHelpersTest.php [hunk 11]
```diff
@@ -490,6 +533,42 @@ function ($argument) {
 	$this->assertStringNotContainsString('--typography-content:', $output);
 });
 
+test('Asserts that outputCssVariables returns variables custom for select type if variable keys is set to object.', function () {
```
Comment (goranalkovic-infinum) [inline tests/Helpers/ComponentHelpersTest.php:536]:
```suggestion
test('Asserts that outputCssVariables returns custom variables for the selected type if variable key is set to object.', function () {
```

File: tests/Helpers/ComponentHelpersTest.php [hunk 12]
```diff
@@ -490,6 +533,42 @@ function ($argument) {
...
+		'uniqueString'
+	);
+
+	$this->assertIsString($output);
+	$this->assertStringContainsString('--test-select-variable-test1: novi1;', $output);
+	$this->assertStringContainsString('--test-select-variable-test2: novi2;', $output);
+	$this->assertStringNotContainsString('--typography-content:', $output);
+});
+
+test('Asserts that outputCssVariables doesn\'t return variable if value is emapty', function () {
```
Comment (goranalkovic-infinum) [inline tests/Helpers/ComponentHelpersTest.php:555]:
```suggestion
test('Asserts that outputCssVariables doesn\'t return a variable if the value is empty', function () {
```

File: tests/Helpers/ComponentHelpersTest.php [hunk 13]
```diff
@@ -490,6 +533,42 @@ function ($argument) {
...
+		$attributes,
+		$manifest,
+		'uniqueString'
+	);
+
+	$this->assertIsString($output);
+	$this->assertStringNotContainsString('--typography-content:', $output);
+});
+
 test('Asserts that outputCssVariables will not return css variables if data is empty', function () {
```
Comment (goranalkovic-infinum) [inline tests/Helpers/ComponentHelpersTest.php:572]:
```suggestion
test('Asserts that outputCssVariables will not return CSS variables if data is empty', function () {
```

File: tests/Helpers/ComponentHelpersTest.php [hunk 14]
```diff
@@ -535,13 +614,12 @@ function ($argument) {
...
 
 	$this->assertIsString($output);
-	$this->assertStringContainsString('--typography-custom: center center;', $output);
 	$this->assertStringContainsString('--typography-custom-horizontal: center;', $output);
 	$this->assertStringContainsString('--typography-custom-vertical: center;', $output);
 	$this->assertStringNotContainsString('--button-content:', $output);
 });
 
-test('aaaAsserts that outputCssVariables returns custom variables for custom type if options key is set.', function () {
+test('Asserts that outputCssVariables doesn\'t return custom variables for custom type if options key is not set.', function () {
```
Comment (goranalkovic-infinum) [inline tests/Helpers/ComponentHelpersTest.php:622]:
```suggestion
test('Asserts that outputCssVariables doesn\'t return custom variables for custom type if the options key is not set.', function () {
```

File: tests/data/src/Blocks/components/typography/manifest.json [hunk 15]
```diff
@@ -79,7 +79,14 @@
 			{
 				"label": "Test 3",
 				"value": "test-3",
-				"variable": "custom-variable-3"
+				"variable": {
+					"test1": "novi1",
```
Comment (goranalkovic-infinum) [inline tests/data/src/Blocks/components/typography/manifest.json:83]: Should be in english

File: src/Helpers/Components.php [hunk 16]
```diff
@@ -258,6 +258,38 @@ public static function checkAttr(string $key, array $attributes, array $manifest
...
+	 * @throws \Exception If missing keyName in responsiveAttributes.
+	 *
+	 * @return mixed
+	 */
+	public static function checkAttrResponsive(string $keyName, array $attributes, array $manifest, string $componentName = '')
+	{
+		$output = [];
+
+		if (!isset($manifest['responsiveAttributes'])) {
+			throw new \Exception("It looks like you are missing responsiveAttributes key in your {$componentName} manifest.");
```
Comment (goranalkovic-infinum) [inline src/Helpers/Components.php:279]:
```suggestion
			throw new \Exception("It looks like you are missing the responsiveAttributes key in your {$componentName} manifest.");
```

File: src/Helpers/Components.php [hunk 17]
```diff
@@ -258,6 +258,38 @@ public static function checkAttr(string $key, array $attributes, array $manifest
...
+		}
+
+		if (!isset($manifest['responsiveAttributes'][$keyName])) {
+			throw new \Exception("It looks like you are missing {$keyName} key in your manifest responsiveAttributes array.");
```
Comment (goranalkovic-infinum) [inline src/Helpers/Components.php:283]:
```suggestion
			throw new \Exception("It looks like you are missing the {$keyName} key in your manifest responsiveAttributes array.");
```

**Review by goranalkovic-infinum — CHANGES_REQUESTED** [review]
Left a couple of suggested changes, nice job!

File: src/Helpers/Components.php [hunk 18]
```diff
@@ -420,6 +451,14 @@ function ($item) use ($attributes, $key) {
 				))[0]['variable'] ?? null;
 
 				$value = $selectVariable === null ? $attributes[$key] : $selectVariable;
+
+				if ($value === "") {
```
Comment (dingo-d) [inline src/Helpers/Components.php:455]:
```suggestion
				if ($value === '') {
```

File: src/Helpers/Components.php [hunk 19]
```diff
@@ -420,6 +451,14 @@ function ($item) use ($attributes, $key) {
...
+					continue;
+				}
+
+				if (gettype($value) === 'array') {
```
Comment (dingo-d) [inline src/Helpers/Components.php:459]:
```suggestion
				if (is_array($value)) {
```

Cleaner, and marginally faster 😄

**Review by goranalkovic-infinum — APPROVED** [review]

**Review by gabriel-glo — APPROVED** [review]

File: src/Helpers/Components.php [hunk 20]
```diff
@@ -126,13 +126,13 @@ public static function render(string $component, array $attributes = [], string
 		// for setting specific styles for components rendered inside other components.
 		if (isset($attributes['parentClass'])) {
 			$component = str_replace('.php', '', $component);
-			printf('<div class="%s">', \esc_attr("{$attributes['parentClass']}__{$component}"));
+			printf(\esc_html('<div class="%s">'), \esc_attr("{$attributes['parentClass']}__{$component}"));
```
Comment (dingo-d) [inline src/Helpers/Components.php:129]: This is a false positive

File: src/Helpers/Components.php [hunk 21]
```diff
@@ -126,13 +126,13 @@ public static function render(string $component, array $attributes = [], string
...
 		}
 
 		require $componentPath;
 
 		if (isset($attributes['parentClass'])) {
-			echo '</div>';
+			echo \esc_html('</div>');
```
Comment (dingo-d) [inline src/Helpers/Components.php:135]: This is also a false positive and shouldn't be escaped.

**Review by dingo-d — CHANGES_REQUESTED** [review]
Looks good but you shouldn't escape strings, that is a false positive.

**Review by gabriel-glo — APPROVED** [review]

**Pull request merged**